### ISSUE TITLE
Strange  "relation" update ... ( imposm3 diff )

### ISSUE THREAD

**Issue by ImreSamu** [omniscale/imposm3#88]
I have found a strange - and different behaviour in the current version.

my test method:
I am compared  the 1 step  "imposm3 import"  output
with  a  2 step import+diff  output ( based on osmium --derive-change )   

the older imposm3 versions -  has no difference ->  has  no problem,  it is OK 
but  the new version - has some diff with relations 

my  test outputs  per different git versions:
```
./output/201511221225_15111ca__diff_tags_should_be_empty.txt 0            <- OK, no diff
./output/201512221010_901b40b__diff_tags_should_be_empty.txt 0            <- OK, no diff
./output/201601151639_bb3d003__diff_tags_should_be_empty.txt 11418          <- CHECK !
./output/201601281348_c033373__diff_tags_should_be_empty.txt 11418          <- CHECK !
```

Basic test files for replicating :  https://github.com/omniscale/imposm3/pull/87 
If I have time .. I will try to create more tests ...

**Comment by ImreSamu** [comment]
I can detect the problem 
When
+ "use_single_id_space": true,
+  AND  diff update relations 

Then
+  the modified osm relations are duplicated ( not deleted )


simple test code
```bash
#!/bin/bash
set -ex
impconnection=postgis://$PGHOST/$PGDATABASE 
mapping_file=single_table_mapping.json
osm_pbf=./build/single_table.pbf
osc_gz=./build/single_table_02.osc.gz
../imposm3 import -mapping $mapping_file -read $osm_pbf -diff -write -overwritecache -deployproduction -connection $impconnection
psql  -c "select id,osm_id,tags from  osm_all where osm_id= -100000000000031101 ;" 
../imposm3 diff -mapping $mapping_file -connection $impconnection $osc_gz
../imposm3 diff -mapping $mapping_file -connection $impconnection $osc_gz
../imposm3 diff -mapping $mapping_file -connection $impconnection $osc_gz
../imposm3 diff -mapping $mapping_file -connection $impconnection $osc_gz
../imposm3 diff -mapping $mapping_file -connection $impconnection $osc_gz
../imposm3 diff -mapping $mapping_file -connection $impconnection $osc_gz
psql   -c "select id,osm_id,tags from  osm_all where osm_id= -100000000000031101 ;" 
```

result:
```log
root@3663cf4f938d:/go/src/github.com/omniscale/imposm3/test# ./test.sh
+ impconnection=postgis://localhost/imposm3dev
+ mapping_file=single_table_mapping.json
+ osm_pbf=./build/single_table.pbf
+ osc_gz=./build/single_table_02.osc.gz
+ ../imposm3 import -mapping single_table_mapping.json -read ./build/single_table.pbf -diff -write -overwritecache -deployproduction -connection postgis://localhost/imposm3dev
[Feb  2 01:52:21] [INFO] removing existing cache /tmp/imposm3
[Feb  2 01:52:21] [INFO] [     0] C:       0/s (11) N:       0/s (3) W:       0/s (12) R:      0/s (1)
[Feb  2 01:52:21] [INFO] Reading OSM data took: 211.302439ms
[Feb  2 01:52:22] [INFO] [     0] C:       0/s ( 0.0%) N:       0/s (100.0%) W:       0/s (100.0%) R:      0/s (100.0%)
[Feb  2 01:52:22] [INFO] Writing OSM data took: 439.70363ms
[Feb  2 01:52:22] [INFO] [PostGIS] Creating generalized tables took: 51.06µs
[Feb  2 01:52:22] [INFO] [PostGIS] Creating OSM id index on osm_all took: 24.352885ms
[Feb  2 01:52:22] [INFO] [PostGIS] Creating geometry index on osm_all took: 5.17802ms
[Feb  2 01:52:22] [INFO] [PostGIS] Creating geometry indices took: 29.889509ms
[Feb  2 01:52:22] [INFO] Importing OSM data took: 470.02737ms
[Feb  2 01:52:22] [INFO] [PostGIS] Rotating osm_all from import -> public -> backup
[Feb  2 01:52:22] [INFO] [PostGIS] backup of osm_all, to backup
[Feb  2 01:52:22] [INFO] [PostGIS] Rotating tables took: 31.414015ms
[Feb  2 01:52:22] [INFO] Imposm took: 843.070645ms
+ psql -c 'select id,osm_id,tags from  osm_all where osm_id= -100000000000031101 ;'
 id |       osm_id        |       tags        
----+---------------------+-------------------
  1 | -100000000000031101 | "building"=>"yes"
(1 row)

+ ../imposm3 diff -mapping single_table_mapping.json -connection postgis://localhost/imposm3dev ./build/single_table_02.osc.gz
[Feb  2 01:52:22] [WARN] [diff] cannot find state file ./build/single_table_02.state.txt
[Feb  2 01:52:22] [INFO] [diff] Parsing changes, updating cache and removing elements took: 1.250528ms
[Feb  2 01:52:22] [INFO] [     0] C:       0/s (0) N:       0/s (0) W:       0/s (0) R:      0/s (1)
[Feb  2 01:52:22] [INFO] [PostGIS] Updating generalized tables took: 68.402µs
[Feb  2 01:52:22] [INFO] [diff] Writing added/modified elements took: 20.386181ms
[Feb  2 01:52:22] [INFO] [diff] Processing ./build/single_table_02.osc.gz took: 26.289205ms
[Feb  2 01:52:22] [INFO] [     0] C:       0/s (0) N:       0/s (0) W:       0/s (2) R:      0/s (1)
+ ../imposm3 diff -mapping single_table_mapping.json -connection postgis://localhost/imposm3dev ./build/single_table_02.osc.gz
[Feb  2 01:52:22] [WARN] [diff] cannot find state file ./build/single_table_02.state.txt
[Feb  2 01:52:22] [INFO] [diff] Parsing changes, updating cache and removing elements took: 1.036952ms
[Feb  2 01:52:22] [INFO] [     0] C:       0/s (0) N:       0/s (0) W:       0/s (0) R:      0/s (1)
[Feb  2 01:52:22] [INFO] [PostGIS] Updating generalized tables took: 40.74µs
[Feb  2 01:52:22] [INFO] [diff] Writing added/modified elements took: 20.213117ms
[Feb  2 01:52:22] [INFO] [diff] Processing ./build/single_table_02.osc.gz took: 26.77366ms
[Feb  2 01:52:22] [INFO] [     0] C:       0/s (0) N:       0/s (0) W:       0/s (2) R:      0/s (1)
+ ../imposm3 diff -mapping single_table_mapping.json -connection postgis://localhost/imposm3dev ./build/single_table_02.osc.gz
[Feb  2 01:52:23] [WARN] [diff] cannot find state file ./build/single_table_02.state.txt
[Feb  2 01:52:23] [INFO] [diff] Parsing changes, updating cache and removing elements took: 1.138601ms
[Feb  2 01:52:23] [INFO] [     0] C:       0/s (0) N:       0/s (0) W:       0/s (0) R:      0/s (1)
[Feb  2 01:52:23] [INFO] [PostGIS] Updating generalized tables took: 48.392µs
[Feb  2 01:52:23] [INFO] [diff] Writing added/modified elements took: 14.359642ms
[Feb  2 01:52:23] [INFO] [diff] Processing ./build/single_table_02.osc.gz took: 20.808351ms
+ ../imposm3 diff -mapping single_table_mapping.json -connection postgis://localhost/imposm3dev ./build/single_table_02.osc.gz
[Feb  2 01:52:23] [WARN] [diff] cannot find state file ./build/single_table_02.state.txt
[Feb  2 01:52:23] [INFO] [diff] Parsing changes, updating cache and removing elements took: 1.490815ms
[Feb  2 01:52:23] [INFO] [     0] C:       0/s (0) N:       0/s (0) W:       0/s (0) R:      0/s (1)
[Feb  2 01:52:23] [INFO] [PostGIS] Updating generalized tables took: 49.456µs
[Feb  2 01:52:23] [INFO] [diff] Writing added/modified elements took: 15.206781ms
[Feb  2 01:52:23] [INFO] [diff] Processing ./build/single_table_02.osc.gz took: 22.581886ms
[Feb  2 01:52:23] [INFO] [     0] C:       0/s (0) N:       0/s (0) W:       0/s (2) R:      0/s (1)
+ ../imposm3 diff -mapping single_table_mapping.json -connection postgis://localhost/imposm3dev ./build/single_table_02.osc.gz
[Feb  2 01:52:24] [WARN] [diff] cannot find state file ./build/single_table_02.state.txt
[Feb  2 01:52:24] [INFO] [diff] Parsing changes, updating cache and removing elements took: 1.028953ms
[Feb  2 01:52:24] [INFO] [     0] C:       0/s (0) N:       0/s (0) W:       0/s (0) R:      0/s (1)
[Feb  2 01:52:24] [INFO] [PostGIS] Updating generalized tables took: 93.202µs
[Feb  2 01:52:24] [INFO] [diff] Writing added/modified elements took: 9.318911ms
[Feb  2 01:52:24] [INFO] [diff] Processing ./build/single_table_02.osc.gz took: 15.022479ms
[Feb  2 01:52:24] [INFO] [     0] C:       0/s (0) N:       0/s (0) W:       0/s (2) R:      0/s (1)
+ ../imposm3 diff -mapping single_table_mapping.json -connection postgis://localhost/imposm3dev ./build/single_table_02.osc.gz
[Feb  2 01:52:24] [WARN] [diff] cannot find state file ./build/single_table_02.state.txt
[Feb  2 01:52:24] [INFO] [diff] Parsing changes, updating cache and removing elements took: 1.211162ms
[Feb  2 01:52:24] [INFO] [     0] C:       0/s (0) N:       0/s (0) W:       0/s (0) R:      0/s (1)
[Feb  2 01:52:24] [INFO] [PostGIS] Updating generalized tables took: 63.953µs
[Feb  2 01:52:24] [INFO] [diff] Writing added/modified elements took: 9.70525ms
[Feb  2 01:52:24] [INFO] [diff] Processing ./build/single_table_02.osc.gz took: 16.333535ms
[Feb  2 01:52:24] [INFO] [     0] C:       0/s (0) N:       0/s (0) W:       0/s (2) R:      0/s (1)
+ psql -c 'select id,osm_id,tags from  osm_all where osm_id= -100000000000031101 ;'
 id |       osm_id        |       tags        
----+---------------------+-------------------
  1 | -100000000000031101 | "building"=>"yes"
 12 | -100000000000031101 | "building"=>"yes"
 13 | -100000000000031101 | "building"=>"yes"
 14 | -100000000000031101 | "building"=>"yes"
 15 | -100000000000031101 | "building"=>"yes"
 16 | -100000000000031101 | "building"=>"yes"
 17 | -100000000000031101 | "building"=>"yes"
(7 rows)

```

where `single_table_02.osc` is : 
```xml
<?xml version='1.0' encoding='UTF-8'?>
<osmChange version="0.6" generator="Osmosis 0.41">
  <modify>
    <relation id="31101" version="1" timestamp="2011-11-11T00:11:11Z">
      <member type="way" ref="31002" role="outer"/>
      <member type="way" ref="31003" role="outer"/>
      <tag k="type" v="multipolygon"/>
      <tag k="building" v="yes"/>
    </relation>
  </modify>
</osmChange>
```


The program code is OK when  `"use_single_id_space": false `

**Comment by ImreSamu** [comment]
maybe this simple  [1 line change ]( 
https://github.com/ImreSamu/imposm3/commit/05c4986dbbab0b596f7f5c27c14ae38efa58c4ff?diff=unified)  solve the problem.

**Comment by olt** [comment]
Yes, that's it. Thanks for reporting this issue!
Fixed in 8949fe2
